### ISSUE TITLE
Qt eventloop interferes with asyncio.call_later in _abort_queues

### ISSUE THREAD

**Issue by jdranczewski** [ipython/ipykernel#1202]
When `kernelbase._abort_queues()` is called, it sets the `self._aborting = True` flag and schedules a future call to the `stop_aborting()` method, which clears the flag. This scheduling is done through `asyncio`: https://github.com/ipython/ipykernel/blob/93a63fb7b8752899ed95118fa35e56f74eedd0c6/ipykernel/kernelbase.py#L1194-L1207
This is done so that the `stop_on_error_timeout` timer is respected. `schedule_stop_aborting()` then puts `stop_aborting()` on the `msg_queue` through `schedule_dispatch`. So the `_aborting` flag is cleared once the asyncio `call_later` fires _and_ the message queue is processed. This works correctly in normal operation.

When a Qt event loop is attached to the kernel through the `%gui qt` magic, this process is disrupted. A minimal example to reproduce is as follows, with each code block representing an `execute_request` or a cell in Jupyter:
```python
%gui qt6
```
```python
1/0
```
An exception is raised, and as `stop_on_error` is True, `kernelbase._abort_queues()` is called.
```python
print("Hello?")
```
This cell is not executed correctly, replying with a `{"status": "aborted"}`, as `schedule_stop_aborting` has not been called yet, so the `_aborting` flag is True.
```python
print("Hello?")
```
This cell executes correctly, printing "Hello?".

A simple fix is calling
```python
self.schedule_dispatch(stop_aborting)
```
directly instead of through `asyncio`, which is how I fixed this on my computer - this means the issue lies with `asyncio` rather than the message queue I think. I appreciate that a more general fix may be needed for use cases that require non-zero `stop_on_error_timeout` though.

I believe this is the root cause behind issues https://github.com/ipython/ipykernel/issues/1066 and https://github.com/jupyter/notebook/issues/6526 - whenever a Qt loop is used, the next cell after an exception is raised fails to run. I raise this as a separate issue as I believe I've found the direct cause of this general problem, but do not understand the Qt loop well enough to implement a general fix myself.

**Comment by blink1073** [comment]
Hi @jdranczewski, I'm not a Qt expert, but from searching around it does look like its event loop and asyncio are incompatible.  There may be something in https://github.com/CabbageDevelopment/qasync/blob/master/qasync/__init__.py that we could pull from, but I don't have the bandwidth to look into it myself.

**Comment by ccordoba12** [comment]
That's correct, the Qt and asyncio event loops are incompatible by default.

@blink1073, would it be ok to add a dependency on qasync to fix this?

**Comment by blink1073** [comment]
I think it might need to be in IPython?  https://github.com/ipython/ipykernel/blob/8672cb0a4454f8090f7c6779bec5b39a3f08b159/ipykernel/eventloops.py#L519

**Comment by jdranczewski** [comment]
Hi all, thank you very much for your help and suggestions so far, it helped me look in the right directions to understand all of this a lot more!

I think the core of the issue is that all of ipykernel's activity is scheduled on a [Tornado/asyncio event loop](https://www.tornadoweb.org/en/stable/ioloop.html), but when the `%gui` magic is used, we introduce another event loop into the mix in the same thread, and have to juggle between them. Now for the most part this works fine - the GUI event loops are designed to hand off back to Tornado in various circumstances. It's just the `call_later` timer that is an edge case - one cannot run a timer in Tornado if the thread is stuck in a separate event loop. The key then I think is to figure out how to exit the GUI event loop at the right time to deal with this, or perhaps integrate the two loops together with `qasync` somehow.

So, once `%gui` is used, a GUI event loop starts and takes over the kernel thread. **How do the GUI event loops exit and go back to Tornado?** This is all defined in https://github.com/ipython/ipykernel/blob/8672cb0a4454f8090f7c6779bec5b39a3f08b159/ipykernel/eventloops.py
* **The Qt loop** uses a zmq socket event to shut itself down and hand over to Tornado: https://github.com/ipython/ipykernel/blob/8672cb0a4454f8090f7c6779bec5b39a3f08b159/ipykernel/eventloops.py#L96-L100 This is an elegant solution that works most of the time, but is susceptible to this bug - the timer is unrelated to the socket and therefore will never interrupt the Qt loop.
* **The Wx, gtk, and osx loops** use a timer within the GUI loop to hand off to Tornado periodically https://github.com/ipython/ipykernel/blob/8672cb0a4454f8090f7c6779bec5b39a3f08b159/ipykernel/eventloops.py#L161 This is less elegant as it causes unnecessary load by switching between the loops basically all the time, but less susceptible to this bug, as the Tornado timer will fire in the next `kernel._poll_interval`.
* **The Tkinter loop** uses a timer on Windows (not susceptible) and a socket event otherwise (theoretically susceptible). I can test this on a Linux system later and see if the issue occurs.
* **The asyncio loop** does mostly nothing, as the Tornado loop should be already running in most cases.

So these GUI loop are occasionally triggered (by a socket event or a timer) to exit. Once they exit, the next entry to the event loop is scheduled through Tornado https://github.com/ipython/ipykernel/blob/8672cb0a4454f8090f7c6779bec5b39a3f08b159/ipykernel/kernelbase.py#L500 Tornado can then process any events that accumulated and enter the GUI event loop again.

**I see a couple of possible solutions then:**
1. **Ignore `stop_on_error_timeout` if a GUI eventloop is engaged**
Arguably the simplest. We cannot trust that a timer will fire accurately if an event loop is engaged, so we don't support the timer use case. Alternatively, we could specifically not support `stop_on_error_timeout` for the Qt loop and the Tkinter loop (on non-Windows devices).
```python
if self.eventloop:
    self.schedule_dispatch(stop_aborting)
else:
    self.io_loop.call_later(self.stop_on_error_timeout, schedule_stop_aborting)
```
From what I can see, this is the only relevant place where `call_later` is used in `ipykernel`, so this support edge case could be an acceptable solution.

2. **Add a timed periodic exit to the Qt (and Tkinter) event loop**
This would make it consistent with the other loops, but less performant, as now we'll have a back-and-forth hand-off between Qt and Tornado every `kernel._poll_interval` that is actually only needed in the specific edge case that the user has a non-zero `stop_on_error_timeout` _and_ an exception has just arisen.

3. **Explicitly set a Qt timer in `_abort_queues` so that the GUI loop exits and the Tornado timer can be processed**
This could be achieved by an optional attribute for the GUI eventloops for example:
```python
self.io_loop.call_later(self.stop_on_error_timeout, schedule_stop_aborting)
if self.eventloop and hasattr(self.eventloop, '_schedule_exit'):
    self.eventloop._schedule_exit(self.stop_on_error_timeout)
```
I think this solution could be a decent common ground,

4. **Only set the timer when `stop_on_error_timeout` is non-zero**
Also a fairly simple fix, but introduces a weird behaviour if the user sets the timeout to non-zero - this error starts happening for seemingly unexplained reasons.
```python
if self.stop_on_error_timeout:
    self.io_loop.call_later(self.stop_on_error_timeout, schedule_stop_aborting)
else:
    self.schedule_dispatch(stop_aborting)
```

5. **Integrate the Qt loop with asyncio/Tornado with `qasync`**
I will admit I don't fully understand if this would be possible, as I'm the least familiar with `asyncio` out of the various tech here. I think a possible problem could be that `qasync` seems to work by setting itself as the main `asyncio` event loop, and that would probably interfere with Tornado? We could potentially implement a _new_ eventloop that inherits from both Tornado's io_loop and `qasync`'s and swap that in when needed. I think @blink1073 is right that it would also need to be a dependency in IPython, because that's where we get Qt from, but it would probably also need to be a dependency here if a custom frankenstein eventloop was to be created. I'm also unsure how this would interfere with users creating their own `QApplication`s, which I do sometimes - `qasync` seems to need its own variant of `QApplication` to work.

**Personally I would be in favour of a combination of 3, and 4,** - only bother with `call_later` if a non-zero `stop_on_error_timeout` is set, and if we need to bother with it, introduce exit timeouts to the GUI event loops affected. Let me know what you think or if I misunderstood or missed anything! I could draft a pull request that achieves this, but preferably only after a consensus is reached on the best way forward. Sorry for the wall of text, but the issue seems to be fairly complex for a seemingly simple bug!

**Comment by blink1073** [comment]
Thanks for the deep dive @jdranczewski!  I agree with your approach, and a PR would be most welcome!

**Comment by jeandet** [comment]
Hello, not sure if this helps, but after two weeks of fight with PySide6/qasync/IPykernel I got [this hack](https://github.com/jeandet/PySideJupyterExperiments/blob/main/QAsync/__init__.py) to work on both Windows and Linux. 
What is weird is that almost all the time it was working on Linux whatever the event loop I tried to force/use even with exceptions thrown while it was almost never working on Windows. On Windows, either the UI or the kernel was frozen. From what I've red here, it sounds like event loops cohabitation kind of work on Linux.

**Comment by jdranczewski** [comment]
Hi, sorry for the delay on my side! I've had to fight with Windows and Linux a bit as well and jump between computers, and figure out that QTimers are not actually precise by default. I finally tested it on Linux properly and it seems to be all good. You can see my suggested changes at https://github.com/ipython/ipykernel/compare/main...jdranczewski:ipykernel:main

I will create a pull request unless you'd like to consider @jeandet 's solution? From what I see it uses qasync to do basically what the Tkinter loop currently does on Windows. I think using timers as I propose may be an easier and more performant solution, as it only comes back to the Tornado loop as needed, but I'm open to discussion of course!
